lib/ex_aws/operation/s3.ex

```diff
@@ -40,21 +41,31 @@ defmodule ExAws.Operation.S3 do
       |> operation.parser.()
     end
 
+    def stream!(%{stream_builder: fun}, config) do
+      fun.(config)
+    end
+
     defp put_content_length_header(headers, "", :get), do: headers
 
     defp put_content_length_header(headers, body, _) do
       Map.put(headers, "content-length", byte_size(body) |> Integer.to_string())
     end
 
-    def stream!(%{stream_builder: fun}, config) do
-      fun.(config)
+    @spec add_bucket_to_path(operation :: ExAws.Operation.S3.t(), config :: map) ::
```

Comment:
What about defining the `operation` type in line 18, so you can reuse it and make the function single-line?

Reply:
Not sure what you mean by this?

Adding the bucket ends up modifying either the host or the path, so it needs to take both and return both. Otherwise, the logic of which needs to be modified ends up outside this function, which is even worse.

Reply:
Sorry let me explain it better. What I mean is:

```elixir
#...

@type t :: %__MODULE__{}
@type operation :: ExAws.Operation.S3.t()
@type config :: map()
#...

@spec add_bucket_to_path(operation, config) :: {operation, config}
def add_bucket_to_path(operation, config \\ %{})

#...
```

Reply:
aaah! yes, i see what you mean. personally, i prefer to see it spelled out fully so i don't have to go chasing definitions (even with the LS it's not great). even though it is more verbose, at a glance one can see exactly what it is doing.